### ISSUE TITLE
Trying to apply a user settings fails with a 400 bad request error 

### ISSUE THREAD

**Issue by Mellow420** [mattermost/mattermost-plugin-google-calendar#66]
When using `/gcal settings` and clicking one of the options nothing happens. When looking into the Dev Console you can see the following request being responded to with a error 400.

When inspecting the request you can see that the `selection_option` is empty.

POST https://mattermost.adyen.com/api/v4/posts/<redacted>/actions/<redacted> 
```json
{"selected_option":""}
```

The server then responds with:

```
{
    "id": "api.post.do_action.action_integration.app_error",
    "message": "Action integration error.",
    "detailed_error": "",
    "request_id": "<redacted>",
    "status_code": 400
}
```

**How to reproduce**

1. Setup the plugin
2. Go through `gcal connect`
3. execute `gcal settings`
4. Click any of the settings
5. Nothing happens. Bad request error in the dev console

Could someone point me to where the code for this functionality lives? I tried looking into it however its not clear to me which part of the code handles this.

**Comment by fmartingr** [comment]
Hey @Mellow420, sorry you are experiencing this!

The code right now lives in a branch under the Microsoft Calendar Plugin: https://github.com/mattermost/mattermost-plugin-mscalendar/pull/267, we are in the process of migrating this into a separate repository entirely.

**Comment by Mellow420** [comment]
Alright, I tried having a look but the fact this is split between repo's makes it quite a pain , and since I know very little about GO in general I am not able to figure out how this works.

Would be greatly appreciated if someone could have a look at this.

Obviously willing to provide more info as needed around this issue/

**Comment by fmartingr** [comment]
Did you have any relevant log messages in your Mattermost instance around the same time the error poped up? We have been using it for a while and didn't receive any report like this.

**Comment by MasinAD** [comment]
I found a solution.

Our Mattermost instance runs in an LXC. Actually, it doesn't matter if it's LXC, nspawn, docker, podman. The important part is it's containerized and behind a reverse proxy, thus not reachable from the outside.

The way the plugin works is by requesting an http connection to the Mattermost instance. It resolves the hostname according to the container's name resolution strategy and receives the external IP address. It then tries to connect to this IP address. But the way iptables and NAT work this doesn't work. I'm not deep enough into netfilter, iptables and stuff, but this hairpinning stuff isn't considered with our setup.

A workaround is to put the reverse proxy's local IP address in the `/etc/hosts` for resolving Mattermost's hostname.

Now the plugin accepts changes to the user settings.

**Comment by fmartingr** [comment]
> I found a solution.
> 
> Our Mattermost instance runs in an LXC. Actually, it doesn't matter if it's LXC, nspawn, docker, podman. The important part is it's containerized and behind a reverse proxy, thus not reachable from the outside.
> 
> The way the plugin works is by requesting an http connection to the Mattermost instance. It resolves the hostname according to the container's name resolution strategy and receives the external IP address. It then tries to connect to this IP address. But the way iptables and NAT work this doesn't work. I'm not deep enough into netfilter, iptables and stuff, but this hairpinning stuff isn't considered with our setup.
> 
> A workaround is to put the reverse proxy's local IP address in the /etc/hosts for resolving Mattermost's hostname.
> 
> Now the plugin accepts changes to the user settings.

Yes, this may be it since the actions call URLs using the SITE_URL set up in Mattermost.

it currently works as such:

```mermaid
sequenceDiagram
    participant Client
    participant Server
    Client->>Server: Clicks on setting button
    Note left of Server: From client `$SITE_URL/api/v4/posts/xxx/action/xxxx`
    Server-->>Server: Calls plugin URL to save setting
    Note left of Server: From server`$SITE_URL/plugins/com.mattermost.gcal/settings/xxx`
    Server-->>Client: Updates post with setting saved
```

You can either solve it the way you did or by adding `A` rules to the DNS server being used by your LXC services to point to your reverse proxy, which is what it is expected, at least for servers exposed to the internet.
